### DISCUSSION TITLE
Add ability to force dynamic resolution for generic methods

### DISCUSSION THREAD

**Discussion by dmccarty-incomm** [dotnet/csharplang#3671] [Language Ideas]
(Note: I have a StackOverflow question about this topic [here](https://stackoverflow.com/q/62843489/1467396))
Consider the following code:

```
public class GenericTest
{
	public void RunTest<T>(T withInput) => Foo(withInput);
	public void Foo<T>(T withInput) => Console.WriteLine("Ran from TryFoo<T>");
	public void Foo(string withInput) => Console.WriteLine("Ran from TryFoo(string)");
}
```
As it is written, there is no way to call the string-specific "Foo" because the only publicly exposed member is RunTest and since it is a generic method, when it calls Foo, it doesn't know when T will be a string and when it will not, so it just always links to the generic "Foo".  Even if I were to run "new GenericTest().RunTest("Hi");" my output would still be "Ran from TryFoo<T>".    Since C# is a staticly bound language by design, this default is correct, even if this specific example might be a bit unexpected.

A commonly suggested work-around for this is to force the runtime to dynamically link to the string-specific Foo by casting the input as a dynamic like so:

	public void RunTest<T>(T withInput) => Foo((dynamic)withInput);

and this works well until the "T" is an out parameter.  If we were to modify the example as follows:
```
public class GenericTest
{
	public void TryRunTest<T>(out T withOutput) => TryFoo(out withOutput);
	
	private void TryFoo<T>(out T withOutput)
	{
		withOutput = default(T);
		Console.WriteLine("Ran from TryFoo<T>");
	}
	private void TryFoo(out string withOutput)
	{
		withOutput = "Strings are special";
		Console.WriteLine("Ran from TryFoo(string)");
	}
}
```
we can no longer cast to dynamic.  If we try
	TryFoo(out (dynamic)withOutput);
we get the error "CS1510 A ref or out value must be an assignable variable" 

It seems like it could be useful to have SOME way to tell the runtime that "when we get to this method, I'd like you to check other overloads to see if there is a better fit." Exactly HOW we go about that doesn't matter much to me, although here are a few suggestions that may be worth considering:

1 Attribute:
	[DynamicResolution]
	private void TryFoo<T>(out T withOutput)

2 Keyword in declaration (I'm using "dynamic", but others may work better) 
This 1st one makes the most sense to me since the dynamic resolution applies to the method as a whole and not to the variables in it per se:

	private dynamic void TryFoo<T>(out T withOutput) 

But it could go in other places, too

	private void TryFoo<dynamic T>(out T withOutput)

	private void TryFoo<T>(dynamic out T withOutput)

If we have multiple types, I think a single "dynamic" at the beginning of the list is all we need/want

	private TResult Bar<dynamic T1, T2, T3, TResult>(T1 input1, T2 input2, out T3 output)

	private TResult Bar<T1, T2, T3, TResult>(dynamic T1 input1, T2 input2, out T3 output)

3 Creating a block similar to "lock", or "using" that contains the overloads we want dynamically linked

	dynamic
	{
	   public void Foo<T>(T withInput) => Console.WriteLine("Ran from TryFoo<T>");
	   public void Foo(string withInput) => Console.WriteLine("Ran from TryFoo(string)");
	}

4 Generic constraints

	private void TryFoo<dynamic T>(out T withOutput) where T: dynamic

I'm sure there are a myriad of other ways to go about it, as well.

**Comment by YairHalberstadt** [comment]
Can you give a realistic example where this would be useful?

**Comment by dmccarty-incomm** [comment]
I have some legacy code that makes heavy use of the following:


```
private static bool TryConvert<T>(string input, out T output)
{
	output = default;
	try
	{
		var converter = System.ComponentModel.TypeDescriptor.GetConverter(typeof(T));
		if (!converter.IsValid(input))
			return false;

		output = (T)converter.ConvertFromString(input);
		return true;
	}
	catch (NotSupportedException)
	{
		return false;
	}
}
```

TryConvert is called from a few other generic methods, which are called from many other places, all of which essentially are going to fetch a string from somewhere and return an int, or a long or a bool (or whatever).  This has worked very well, but a new requirement is to allow a string that will be converted into a DirectoryInfo.  When we try to use the framework as built, the line

	if (!converter.IsValid(input))

fails and returns false.  The only way to satisfy this new requirement without having to do a ton of refactoring is to change this common method.  Originally I had just tried adding this overload to solve the problem:

	private static bool TryConvert(string input, out DirectoryInfo output)
	{
		try
		{
			output = new DirectoryInfo(input);
			return true;
		}
		catch
		{
			output = null;
			return false;
		}
	}

but that never gets called, as described in the original post.  I can force it to use this overload as follows:

	if (!converter.IsValid(input))
	{
		if (typeof(T) == typeof(DirectoryInfo))
		{
			var retval = TryConvert(input, out DirectoryInfo di);
			output = (T)(dynamic)di;
			return retval;
		}

		return false;
	}

but not only is that ugly and hacky, but since it requires specifically checking for when T is a DirectoryInfo, the overload really isn't buying us anything at all.  I may as well just do the logic in TryConvert:

	if (!converter.IsValid(input))
	{
		if (typeof(T) == typeof(DirectoryInfo))
		{
			try
			{
				output = (T)(dynamic) new DirectoryInfo(input);
				return true;
			}
			catch
			{
				return false;
			}
		}

		return false;
	}

If all we ever have to deal with is DirectoryInfo, that's fine, but what if we start getting other types we have to deal with?  This could start to get long and ugly.  At that point, I'll be forced to refactor everything.  And that's fine--that's part of life with code, but at the same time, being able to opt-in to dynamic linking could be a nice addition to the language.

**Comment by HaloFour** [comment]
Overload resolution happens at compile time based on the static types known to the compiler.  There really isn't a mechanism to move overload resolution to runtime short of writing dynamic code which has a built-in binder that tries to duplicate the logic of the compiler.  I don't think it would be wise to make runtime resolution and dispatch a part of the language proper.

IMO a better solution would be to modify this `TryConvert` method so that you can register a dictionary of additional converters for known types.

**Comment by svick** [comment]
> we can no longer cast to dynamic. If we try
> TryFoo(out (dynamic)withOutput);
> we get the error "CS1510 A ref or out value must be an assignable variable"

You can't cast the `out` parameter to `dynamic` like that, but you can cast `this`:

```c#
((dynamic)this).TryFoo(out withOutput);
```

(Though this approach only works for instance methods.)

**Comment by dmccarty-incomm** [comment]
@svick I'm lucky because TryConvert was declared as static, but doesn't need to be.  I've updated my code based on your suggestion and this works very nicely:

```
private bool TryConvert<T>(string input, out T output)
{
	output = default;
	try
	{
		var converter = System.ComponentModel.TypeDescriptor.GetConverter(typeof(T));
		if (!converter.IsValid(input))
			return ((dynamic)this).TryConvertSpecial(input, out output);

		output = (T)converter.ConvertFromString(input);
		return true;
	}
	catch (NotSupportedException)
	{
		return false;
	}
}

private bool TryConvertSpecial(string input, out DirectoryInfo output)
{
	try
	{
		output = new DirectoryInfo(input);
		return true;
	}
	catch
	{
		output = null;
		return false;
	}
}

private bool TryConvertSpecial<T>(string input, out T output)
{
	output = default;
	return false;
}
```

@HaloFour I'm sure you know much more about it than I do, but since it appears we are already (at least to some extent) doing overload resolution at runtime when we do

	Foo((dynamic)withInput);

or

	((dynamic)this).TryFoo(out withOutput);

I was hoping that we could leverage that same thing, but just in a more declarative way (if that's the right way of saying it).  That may qualify as being just syntactic sugar and perhaps if you find yourself in a position where this would be useful, it simply means you didn't write your code right in the first place and so maybe it's just a bad idea.  Either way, thanks to you and everyone else for considering it.
